### PR TITLE
Allow custom overrites for package resolving and optional sys.path support

### PR DESCRIPTION
**Pull request by barteltm**
# Motivation

Certain build systems or configurations have Python packages located in a non-default path. Those are not found by _Codegen_ per default.

# Content

_Codegen_ now considers the `PYTHONPATH` environment variable when resolving packages.

# Testing

Added unit tests which resolve packages only when the `PYTHONPATH` is set in a correct way.

# Please check the following before marking your PR as ready for review

- [x] I have added tests for my changes
- [x] I have updated the documentation or added new documentation as needed

### CODE REVIEW

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -51,6 +51,10 @@ uv run pytest tests/unit -n auto
 uv run pytest tests/integration/codemod/test_codemods.py -n auto
 ```
 
+> [!TIP]
+>
+> - If on Linux the error `OSError: [Errno 24] Too many open files` appears then increase it with `ulimit -n 10000`
```
Comment (tkfoss) [inline CONTRIBUTING.md]: Helpful tip but increasing ulimit has implications, I would perhaps reword this to a hint rather than operation - "... then increase it with `ulimit -n 10000`" --> "... then you might want to increase your ulimit"

Comment (barteltm) [reply CONTRIBUTING.md]: Good catch. Changed.

File: src/codegen/sdk/python/import_resolution.py [hunk 2]
```diff
@@ -148,6 +157,26 @@ def resolve_import(self, base_path: str | None = None, *, add_module_name: str |
         # ext = ExternalModule.from_import(self)
         # return ImportResolution(symbol=ext)
 
+    @noapidoc
+    @reader
+    def _file_by_pythonpath(self, filepath: str) -> SourceFile | None:
+        """Helper to check if a certain import is in the list of files when a PYTHONPATH
+        is set. Returns either None or the SourceFile.
+        """
+        python_paths: str | None = os.environ.get("PYTHONPATH", None)
```
Comment (eacodegen) [inline src/codegen/sdk/python/import_resolution.py]: Can we use sys.path instead of PYTHONPATH?

**Review by eacodegen — CHANGES_REQUESTED** [review]
Would it be possible for you to add two options:
1. Make respecting sys.path opt in
2. Add additional overrides 

Example: if codegen is installed a virtual environment, we wouldn't want to make the user set PYTHONPATH/install it into the project environment. We'd also not want to accidentally include code from their virtual enviornment

**Review by eacodegen — COMMENTED** [review]
Thanks for the PR by the way! Let us know if you ever have any questions.

File: tests/unit/codegen/sdk/python/import_resolution/test_import_resolution.py [hunk 3]
```diff
@@ -178,7 +178,7 @@ def test_import_properties_multi_imports(tmpdir) -> None:
         assert z[7].alias.source == "A"
 
 
-def test_import_resolution_file_import(tmpdir: str) -> None:
+def test_import_resolution_file_import(tmpdir: str, monkeypatch) -> None:
```
Comment (eacodegen) [inline tests/unit/codegen/sdk/python/import_resolution/test_import_resolution.py]: I would also add more tests, preferably several separate functions, to catch various edge cases

**Review by eacodegen — CHANGES_REQUESTED** [review]

File: src/codegen/sdk/python/import_resolution.py [hunk 4]
```diff
@@ -107,6 +108,13 @@ def resolve_import(self, base_path: str | None = None, *, add_module_name: str |
         if file := self.ctx.get_file(filepath):
             return ImportResolution(from_file=file, symbol=None, imports_file=True)
 
+        # =====[ Check if we are importing an entire file with custom resolve path or sys.path enabled ]=====
+        if len(self.ctx.config.py_resolve_overrides) > 0 or self.ctx.config.py_resolve_syspath:
+            # Handle resolve overrides first if both is set
+            resolve_paths: list[str] = self.ctx.config.py_resolve_overrides + (sys.path if self.ctx.config.py_resolve_syspath else [])
```
Comment (bagel897) [inline src/codegen/sdk/python/import_resolution.py]: Can we refactor this to it's own function?

File: src/codegen/sdk/python/import_resolution.py [hunk 5]
```diff
@@ -107,6 +108,13 @@ def resolve_import(self, base_path: str | None = None, *, add_module_name: str |
         if file := self.ctx.get_file(filepath):
```
Comment (bagel897) [inline src/codegen/sdk/python/import_resolution.py]: If we do have a sys.path/override configured, can we make sure to use that path to resolve imports in case of conflicts?  Like if we have  a/b.py and b.py and a path of ["a", ""], we should resolve the import from a  if sys path/overrides are enabled.

**Review by bagel897 — CHANGES_REQUESTED** [review]

File: src/codegen/configs/models/codebase.py [hunk 6]
```diff
@@ -17,6 +17,8 @@ def __init__(self, prefix: str = "CODEBASE", *args, **kwargs) -> None:
     disable_graph: bool = False
     generics: bool = True
     import_resolution_overrides: dict[str, str] = Field(default_factory=lambda: {})
+    py_resolve_overrides: list[str] = Field(default_factory=lambda: [])
```
Comment (EdwardJXLi) [inline src/codegen/configs/models/codebase.py]: The naming for `py_resolve_overrides` might be little confusing with `import_resolution_overrides`? 

`py_resolve_overrides` means alternate paths to look for during import resolution, while `import_resolution_overrides` means actually doing a string replace on the import during resolution.

File: src/codegen/sdk/python/import_resolution.py (hunk 4 above)
Comment (EdwardJXLi) [reply src/codegen/sdk/python/import_resolution.py]: I think its fine as is?

File: src/codegen/sdk/python/import_resolution.py [hunk 7]
```diff
@@ -148,6 +164,20 @@ def resolve_import(self, base_path: str | None = None, *, add_module_name: str |
         # ext = ExternalModule.from_import(self)
         # return ImportResolution(symbol=ext)
 
+    @noapidoc
+    @reader
+    def _file_by_custom_resolve_paths(self, resolve_paths: list[str], filepath: str) -> SourceFile | None:
```
Comment (EdwardJXLi) [inline src/codegen/sdk/python/import_resolution.py:186]: We could honestly have this as a feature for both TS and PY.

File: src/codegen/configs/models/codebase.py (hunk 6 above)
Comment (EdwardJXLi) [reply src/codegen/configs/models/codebase.py]: Maybe like `import_resolution_additional_paths` or just `import_resolution_paths`? This also makes it so that we can add this for TypeScript support as well.

**Review by EdwardJXLi — CHANGES_REQUESTED** [review]

File: src/codegen/sdk/python/import_resolution.py (hunk 7 above)
Comment (bagel897) [reply src/codegen/sdk/python/import_resolution.py:186]: Down to add that down the line!

File: src/codegen/configs/models/codebase.py (hunk 6 above)
Comment (barteltm) [reply src/codegen/configs/models/codebase.py]: Changed it to `import_resolution_paths` as it's in my option clear enough.

File: src/codegen/sdk/python/import_resolution.py (hunk 5 above)
Comment (barteltm) [reply src/codegen/sdk/python/import_resolution.py]: Changed the code order and added a dedicated test.

File: src/codegen/sdk/python/import_resolution.py (hunk 4 above)
Comment (barteltm) [reply src/codegen/sdk/python/import_resolution.py]: Both would be perfectly fine for me.

**Review by bagel897 — APPROVED** [review]

**Review by EdwardJXLi — APPROVED** [review]

**Comment by tkfoss** [comment]
thank you @spezifant
